sys-utils/libcmt/src/keccak.c

```diff
@@ -0,0 +1,224 @@
+/* Copyright Cartesi and individual authors (see AUTHORS)
+ * SPDX-License-Identifier: Apache-2.0
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ * http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+#include "keccak.h"
+#include "abi.h"
+
+#include <string.h>
+
+// Helper macros for stringification
+#define TO_STRING_HELPER(X) #X
+#define TO_STRING(X) TO_STRING_HELPER(X)
+
+// Define loop unrolling depending on the compiler
+#if defined(__clang__)
+#define UNROLL_LOOP(n) _Pragma(TO_STRING(unroll(n)))
+#elif defined(__GNUC__) && !defined(__clang__)
+#define UNROLL_LOOP(n) _Pragma(TO_STRING(GCC unroll(n)))
+#else
+#define UNROLL_LOOP(n)
+#endif
+
+#define ROTL64(x, y) (((x) << (y)) | ((x) >> (64 - (y))))
+
+/** Initialize a keccak state
+ * @note don't port. use @ref cmt_keccak_init */
+#define CMT_KECCAK_INIT(STATE)                                                                                         \
```

Comment:
Clang tidy made a mess here, I'll change it back and add a comment so it doesn't touch it.